File: automator/src/main/java/com/stardust/automator/filter/DfsFilter.java
package com.stardust.automator.filter;

import com.stardust.automator.UiObject;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Stardust on 2017/3/9.
 */

public abstract class DfsFilter implements ListFilter, Filter {

    @Override
    public List<UiObject> filter(List<UiObject> nodes) {
        ArrayList<UiObject> list = new ArrayList<>();
        for (UiObject node : nodes) {
            if (isIncluded(node)) {
                list.add(node);
            }
            filterChildren(node, list);
        }
        return list;
    }

    public List<UiObject> filter(UiObject node) {
        ArrayList<UiObject> list = new ArrayList<>();
        if (isIncluded(node)) {
            list.add(node);
        }
        filterChildren(node, list);
        return list;
    }

    private void filterChildren(UiObject parent, List<UiObject> list) {
        for (int i = 0; i < parent.getChildCount(); i++) {
            UiObject child = parent.child(i);
            if (child == null)
                continue;
            boolean included = isIncluded(child);
            if (included) {
                list.add(child);
            }
            filterChildren(child, list);
            if (!included) {
                child.recycle();
            }
        }
    }

    protected abstract boolean isIncluded(UiObject nodeInfo);
}

File: automator/src/main/java/com/stardust/automator/filter/DfsFilter.kt
package com.stardust.automator.filter

import com.stardust.automator.UiObject

import java.util.ArrayList

/**
 * Created by Stardust on 2017/3/9.
 */

abstract class DfsFilter : ListFilter(), Filter {

    override fun filter(nodes: List<UiObject>): List<UiObject> {
        val list = ArrayList<UiObject>()
        for (node in nodes) {
            if (isIncluded(node)) {
                list.add(node)
            }
            filterChildren(node, list)
        }
        return list
    }

    override fun filter(node: UiObject): List<UiObject> {
        val list = ArrayList<UiObject>()
        if (isIncluded(node)) {
            list.add(node)
        }
        filterChildren(node, list)
        return list
    }

    private fun filterChildren(parent: UiObject, list: MutableList<UiObject>) {
        for (i in 0 until parent.childCount) {
            val child = parent.child(i) ?: continue
            val included = isIncluded(child)
            if (included) {
                list.add(child)
            }
            filterChildren(child, list)
            if (!included) {
                child.recycle()
            }
        }
    }

    protected abstract fun isIncluded(nodeInfo: UiObject): Boolean
}
